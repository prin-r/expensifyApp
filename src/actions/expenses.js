import uuid from 'uuid';

// add expense
export const addExpense = (
    {
        description = '',
        note = '',
        amount = 0,
        createAt = 0
    } = {}
) => ({
    type: 'ADDEXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createAt
    }
});
// remove expense
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});
// edit expense
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});