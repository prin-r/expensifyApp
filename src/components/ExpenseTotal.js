import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenseTotal from '../selectors/expenseTotal';

export const ExpenseTotal = ({ expenseCount, expenseTotal }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    return (
        <div>
            <h1>Viewing {expenseCount} {expenseWord} totaling {numeral(expenseTotal / 100).format('$0,0.00')}</h1>
        </div>
    );
};

const mapStateToProps = ({ expenses }) => {
    return selectExpenseTotal(expenses);
};

export default connect(mapStateToProps)(ExpenseTotal);