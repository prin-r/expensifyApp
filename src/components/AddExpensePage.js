import React from 'react';
import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expenses';
import { connect } from 'react-redux';

const AddExpensePage = ({ dispatch, history }) => (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm 
            onSubmit={(expense) => {
                dispatch(addExpense(expense));
                history.push('/');
            }}
        />
    </div>
);

export default connect()(AddExpensePage);