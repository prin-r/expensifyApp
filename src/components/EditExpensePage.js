import React from 'react';
import ExpenseForm from './ExpenseForm'
import { editExpense } from '../actions/expenses';
import { removeExpense } from '../actions/expenses';
import { connect } from 'react-redux';

const EditExpensePage = ({ expense, dispatch, history, match }) => {
    return (
        <div>
            Editing the expense with id of {match.params.id}
            <ExpenseForm
                expense={expense}
                onSubmit={(expense) => {
                    dispatch(editExpense(match.params.id, expense));
                    history.push('/');
                }}
            />
            <button onClick={() => {
                dispatch(removeExpense({ id: match.params.id }));
                history.push('/');
            }}>remove</button>
        </div>
    );
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(expense => expense.id === props.match.params.id)
    };
};

export default connect(mapStateToProps)(EditExpensePage);