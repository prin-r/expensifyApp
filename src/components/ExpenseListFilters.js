import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

const setTextFilterLocal = (e,props) => {
    props.dispatch(setTextFilter(e.target.value));
    console.log(e.target.value);
};

const setSortBy = (e,props) => {
    switch (e.target.value) {
        case 'date':
            props.dispatch(sortByDate());
            break;
        case 'amount':
            props.dispatch(sortByAmount());
            break;
        default:
            break;
    };
    console.log(e.target.value);
}

class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={(e) => setTextFilterLocal(e,this.props)} />
                <select value={this.props.filters.sortBy} onChange={(e) => setSortBy(e,this.props)}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        );
    };
};

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const ConnectedExpenseListFilters = connect(mapStateToProps)(ExpenseListFilters);

export default ConnectedExpenseListFilters;