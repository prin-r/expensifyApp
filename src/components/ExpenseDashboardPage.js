import React from 'react';
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseTotal from './ExpenseTotal';

const ExpenseDashboardPage = ({ history }) => (
    <div>
        <ExpenseTotal />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;
