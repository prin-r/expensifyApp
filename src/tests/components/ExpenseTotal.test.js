import React from 'react';
import expenses from '../fixtures/expenses';
import { ExpenseTotal } from '../../components/ExpenseTotal';
import { shallow } from 'enzyme';

test('test total expense component with 1 expense', () => {
    const wrapper = shallow(<ExpenseTotal expenseCount={1} expenseTotal={999} /> );
    expect(wrapper).toMatchSnapshot();
});

test('test total expense component with multiple expenses', () => {
    const wrapper = shallow(<ExpenseTotal expenseCount={3} expenseTotal={3478409023858} /> );
    expect(wrapper).toMatchSnapshot();
});