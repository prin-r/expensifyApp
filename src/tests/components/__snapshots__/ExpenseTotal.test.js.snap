// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test total expense component with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totaling 
    $9.99
  </h1>
</div>
`;

exports[`test total expense component with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totaling 
    $34,784,090,238.58
  </h1>
</div>
`;
